conjure-go-client/httpclient/client.go

```diff
@@ -128,20 +127,33 @@ func (c *clientImpl) Do(ctx context.Context, params ...RequestParam) (*http.Resp
 	return resp, werror.Error("could not find live server")
 }
 
-// If lastURI was already marked failed, we perform a backoff as determined by the retrier.
-// Otherwise, we add lastURI to failedURIs and return the next URI and its offset immediately
+// If lastURI was already marked failed, we perform a backoff as determined by the retrier before returning the next URI and its offset.
+// Otherwise, we add lastURI to failedURIs and return the next URI and its offset immediately.
 func nextURIOrBackoff(lastURI string, uris []string, offset int, failedURIs map[string]struct{}, retrier retry.Retrier) (nextURI string, nextURIOffset int) {
 	_, performBackoff := failedURIs[lastURI]
-	failedURIs[lastURI] = struct{}{}
-	nextURIOffset = (offset + 1) % len(uris)
-	nextURI = uris[nextURIOffset]
+	nextURIOffset, nextURI = markFailedAndGetNextURI(failedURIs, lastURI, offset, uris)
 	// If the URI has failed before, perform a backoff
 	if performBackoff {
 		retrier.Next()
 	}
 	return nextURI, nextURIOffset
 }
 
+// Marks the current URI as failed, gets the next URI, and performs a backoff as determined by the retrier.
+func nextURIAndBackoff(lastURI string, uris []string, offset int, failedURIs map[string]struct{}, retrier retry.Retrier) (nextURI string, nextURIOffset int) {
```

Comment:
comment from the peanut gallery that this code could be easier to read if we had a proper url iterator
some kind of object that tracked the state of the url we were using and the failed urls as well
would also be more testable

Reply:
agreed but would describe that as slightly orthogonal?